src/ferny/ssh_askpass.py

```diff
@@ -0,0 +1,181 @@
+    "%{pkcs11_id}": r"(?P<pkcs11_id>.+)",
+    "%{username}": r"(?P<username>[^ @']+)",
+}
+
+
+class PasswordPrompt(SSHAskpassPrompt):
+    PATTERN = r"%{username}@%{hostname}'s password: "
+    username: Optional[str] = None
+    hostname: Optional[str] = None
+
+    async def dispatch(self, responder: 'SshAskpassResponder') -> Optional[str]:
+        return await responder.do_password_prompt(self)
+
+
+class PassphrasePrompt(SSHAskpassPrompt):
+    PATTERN = r"Enter passphrase for key '%{filename}': "
+    filename: str
+
+    async def dispatch(self, responder: 'SshAskpassResponder') -> Optional[str]:
+        return await responder.do_passphrase_prompt(self)
+
+
+class FIDOPINPrompt(SSHAskpassPrompt):
+    PATTERN = r"Enter PIN for %{algorithm} key %{filename}: "
+    algorithm: str
+    filename: str
+
+    async def dispatch(self, responder: 'SshAskpassResponder') -> Optional[str]:
+        return await responder.do_fido_pin_prompt(self)
+
+
+class FIDOUserPresencePrompt(SSHAskpassPrompt):
+    PATTERN = r"Confirm user presence for key %{algorithm} %{fingerprint}\n"
+    ANSWERS = ()
+    algorithm: str
+    fingerprint: str
+
+    async def dispatch(self, responder: 'SshAskpassResponder') -> Optional[str]:
+        return await responder.do_fido_user_presence_prompt(self)
+
+
+class PKCS11PINPrompt(SSHAskpassPrompt):
+    PATTERN = r"Enter PIN for '%{pkcs11_id}: "
+    pkcs11_id: str
+
+    async def dispatch(self, responder: 'SshAskpassResponder') -> Optional[str]:
+        return await responder.do_pkcs11_pin_prompt(self)
+
+
+class HostKeyPrompt(SSHAskpassPrompt):
+    PATTERN = r"Are you sure you want to continue connecting \(yes/no(/\[fingerprint\])?\)\? "
+    EXTRA_PATTERNS = [
+        r"%(fingerprint)[.]$",
+        r"^%{algorithm} key fingerprint is",
+        r"^The fingerprint for the %{algorithm} key sent by the remote host is$"
+    ]
+    ANSWERS = ('yes', 'no')
+    algorithm: str
+    fingerprint: str
+
+    async def dispatch(self, responder: 'SshAskpassResponder') -> Optional[str]:
+        return await responder.do_host_key_prompt(self)
+
+
+def with_helpers(pattern: str) -> str:
+    for name, helper in HELPERS.items():
+        pattern = pattern.replace(name, helper)
+
+    assert '%{' not in pattern
+    return pattern
+
+
+def categorize_ssh_prompt(string: str, stderr: str) -> AskpassPrompt:
+    classes = [
+        FIDOPINPrompt,
+        FIDOUserPresencePrompt,
+        HostKeyPrompt,
+        PKCS11PINPrompt,
+        PassphrasePrompt,
+        PasswordPrompt,
+    ]
+
+    # The last line is the line after the last newline character, excluding the
+    # optional final newline character.  eg: "x\ny\nLAST\n" or "x\ny\nLAST"
+    second_last_newline = string.rfind('\n', 0, -1)
+    if second_last_newline >= 0:
+        last_line = string[second_last_newline + 1:]
+        extras = string[:second_last_newline + 1]
+    else:
+        last_line = string
+        extras = ''
```

Comment:
Is this dance really necessary? As opposed to looking at `string.rstrip()`?

Reply:
As in `string.rstrip().rfind('\n')`?  Indeed, that would probably work... but we'd still need to consider the two separate cases of if we have a second last newline and if we don't...

We may or may not have a trailing newline at the very end, and there are examples in the unit test of both cases... in particular, the "Confirm users presence" should be treated as the prompt, and has a trailing newline, but the other prompts usually end with `": "` and no newline...